### PR TITLE
Instance Group Storage Refactor

### PR DESCRIPTION
**Pull request by nv-kmcgill53**
This refactor is the first step for being able to manipulate instance groups without the need to completely unload and reload the model.

### CODE REVIEW

File: src/backend_model.h [hunk 1]
```diff
@@ -123,8 +124,8 @@ class TritonModel : public Model {
   std::shared_ptr<TritonBackend> backend_;
 
   // The model instances for this model.
```
Comment (dyastremsky) [inline src/backend_model.h:126]: Unrelated to this ticket, but general quality-of-life: it'd be great to clarify in this comment (or in AddInstance) what passive means  here. No need to update it here, but if you're working on these files, would be awesome to clarify it in this or a future PR.

File: src/sequence_batch_scheduler.cc [hunk 2]
```diff
@@ -57,7 +57,11 @@ SequenceBatchScheduler::Create(
              << " backlog queued requests...";
   }
 
-  auto instance_count = model->Instances().size();
+  uint64_t instance_count = 0;
+  for (const auto& pair : model->InstanceGroups()) {
+    instance_count += pair.second.size();
+  }
+  // auto instance_count = model->Instances().size();
```
Comment (dyastremsky) [inline src/sequence_batch_scheduler.cc:64]: Did you mean to remove this?

**Review by dyastremsky — CHANGES_REQUESTED** [review]

File: src/sequence_batch_scheduler.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (nv-kmcgill53) [reply src/sequence_batch_scheduler.cc]: I did, removed.

File: src/backend_model.h (hunk 1 above)
Comment (nv-kmcgill53) [reply src/backend_model.h]: Updated the comment (hopefully) clarifying the storage container and the definition of the passive instance groups.

**Review by dyastremsky — DISMISSED** [review]

**Comment by nv-kmcgill53** [comment]
Note: `filesystem.h` and `model_config_utils.h` only have clang formatting changes.

File: src/backend_model.h [hunk 3]
```diff
@@ -122,9 +123,12 @@ class TritonModel : public Model {
...
 
-  // The model instances for this model.
-  std::vector<std::unique_ptr<TritonModelInstance>> instances_;
-  std::vector<std::unique_ptr<TritonModelInstance>> passive_instances_;
+  // The model instances for this model stored by the
+  // instance groups defined in the model configuration.
+  // Passive instance groups are those instances which are
+  // loaded but not added to the scheduler.
+  std::map<std::string, TritonInstanceGroup> instance_group_map_;
+  std::map<std::string, TritonInstanceGroup> passive_instance_group_map_;
```
Comment (rmccorm4) [inline src/backend_model.h:131]: Do the `map`s need to be ordered?

File: src/backend_model.h [hunk 4]
```diff
@@ -65,9 +66,9 @@ class TritonModel : public Model {
       const uint32_t config_version,
       TRITONSERVER_Message* updated_config_message);
   const std::shared_ptr<TritonBackend>& Backend() const { return backend_; }
-  const std::vector<std::unique_ptr<TritonModelInstance>>& Instances() const
+  const std::map<std::string, TritonInstanceGroup>& InstanceGroups() const
```
[the remark this answers is not in the document]
Comment (nv-kmcgill53) [reply src/backend_model.h:69]: This map doesn't need to be ordered. It ends up being mostly constant once the model is loaded so we mostly read from it. In the context of hot instance loading/unloading, for now the group names will not change. We will assert the instance groups are in the same order and identical as in the previous model config (except for the `count`). Therefore, we will not have any insertions/deletions into the map. All that said, you are right in that we should not incur debt where not needed and I will fix to `unordered_map`.

File: src/backend_model.h (hunk 3 above)
Comment (nv-kmcgill53) [reply src/backend_model.h:131]: See [this comment](https://github.com/triton-inference-server/core/pull/153#discussion_r1063758281)

**Review by rmccorm4 — APPROVED** [review]

**Review by dyastremsky — APPROVED** [review]

**Pull request merged**